kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java

```diff
@@ -105,6 +114,229 @@ void shouldEncryptRecordValue() {
+        assertThat(encryptedFuture).succeedsWithin(Duration.ZERO);
+        MemoryRecords encrypted = encryptedFuture.join();
+        record.value().rewind();
+
+        assertThat(encrypted.batches()).hasSize(2);
+        List<MutableRecordBatch> batches = StreamSupport.stream(encrypted.batches().spliterator(), false).toList();
+        MutableRecordBatch first = batches.get(0);
+        assertThat(first.compressionType()).isEqualTo(CompressionType.NONE);
+        assertThat(first.timestampType()).isEqualTo(TimestampType.CREATE_TIME);
+        assertThat(first.baseOffset()).isEqualTo(1L);
+        assertThat(first).hasSize(1);
+
+        MutableRecordBatch second = batches.get(1);
+        // should we keep the client's compression type?
+        assertThat(second.compressionType()).isEqualTo(emptyBatch.compressionType());
+        assertThat(second.timestampType()).isEqualTo(emptyBatch.timestampType());
+        assertThat(second.baseOffset()).isEqualTo(emptyBatch.baseOffset());
+        assertThat(second).hasSize(0);
+
+        CompletableFuture<MemoryRecords> decryptedFuture = km.decrypt("topic", 1, encrypted, ByteBufferOutputStream::new).toCompletableFuture();
+        assertThat(decryptedFuture).succeedsWithin(Duration.ZERO);
+        MemoryRecords decrypted = decryptedFuture.join();
+
+        assertThat(decrypted.batches()).hasSize(2);
+        List<MutableRecordBatch> decryptedBatches = StreamSupport.stream(decrypted.batches().spliterator(), false).toList();
+        MutableRecordBatch firstDecrypted = decryptedBatches.get(0);
+        assertThat(firstDecrypted.compressionType()).isEqualTo(CompressionType.NONE);
+        assertThat(firstDecrypted.timestampType()).isEqualTo(TimestampType.CREATE_TIME);
+        assertThat(firstDecrypted.baseOffset()).isEqualTo(1L);
+        assertThat(firstDecrypted).hasSize(1);
+        assertThat(firstDecrypted.iterator())
+                .toIterable()
+                .singleElement()
+                .extracting(RecordTestUtils::recordValueAsBytes)
+                .isEqualTo(value);
+
+        MutableRecordBatch secondDecrypted = decryptedBatches.get(1);
+        assertThat(secondDecrypted.compressionType()).isEqualTo(emptyBatch.compressionType());
+        assertThat(secondDecrypted.timestampType()).isEqualTo(emptyBatch.timestampType());
+        assertThat(secondDecrypted.baseOffset()).isEqualTo(emptyBatch.baseOffset());
+    }
+
+    private static RecordBatch controlBatch(int baseOffset, byte[] arbitraryValue) {
+        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(ByteBuffer.allocate(1000), RecordBatch.CURRENT_MAGIC_VALUE, CompressionType.NONE,
+                TimestampType.CREATE_TIME, baseOffset, 1L, 1L, (short) 1, 1, false, true, 1, 1);
+        byte[] key = { 0, 0, (byte) ControlRecordType.ABORT.type(), (byte) (ControlRecordType.ABORT.type() >> 8) };
```

Comment:
I didn't spot the method to do this but it's bound to be in the kafka-clients API, we need a key with a specific schema.